src/interfaces/chain.cpp

```diff
@@ -0,0 +1,419 @@
+        std::string errString;
+        if (!::mempool.CalculateMemPoolAncestors(entry, setAncestors, nLimitAncestors, nLimitAncestorSize,
+                nLimitDescendants, nLimitDescendantSize, errString)) {
+            return false;
+        }
+        return true;
+    }
+    CFeeRate estimateSmartFee(int num_blocks, bool conservative, FeeCalculation* calc) override
+    {
+        return ::feeEstimator.estimateSmartFee(num_blocks, calc, conservative);
+    }
+    int estimateMaxBlocks() override
+    {
+        return ::feeEstimator.HighestTargetTracked(FeeEstimateHorizon::LONG_HALFLIFE);
+    }
+    CFeeRate poolMinFee() override
+    {
+        return ::mempool.GetMinFee(gArgs.GetArg("-maxmempool", DEFAULT_MAX_MEMPOOL_SIZE) * 1000000);
+    }
+    bool getPruneMode() override { return ::fPruneMode; }
+    bool p2pEnabled() override { return g_connman != nullptr; }
+    int64_t getAdjustedTime() override { return GetAdjustedTime(); }
+    void initMessage(const std::string& message) override { ::uiInterface.InitMessage(message); }
+    void initWarning(const std::string& message) override { InitWarning(message); }
+    void initError(const std::string& message) override { InitError(message); }
+    UniValue generateBlocks(std::shared_ptr<CReserveScript> coinbase_script,
+        int num_blocks,
+        uint64_t max_tries,
+        bool keep_script) override
+    {
+        return ::generateBlocks(coinbase_script, num_blocks, max_tries, keep_script);
+    }
+    unsigned int parseConfirmTarget(const UniValue& value) override { return ParseConfirmTarget(value); }
+    std::unique_ptr<Handler> handleNotifications(Notifications& notifications) override
+    {
+        return MakeUnique<HandlerImpl>(notifications);
+    }
+    void waitForNotifications() override { SyncWithValidationInterfaceQueue(); }
+    std::unique_ptr<Handler> handleRpc(const CRPCCommand& command) override;
+
+    //! Map of RPC method name to forwarder. If multiple clients provide
+    //! implementations of the same RPC method, the RPC forwarder can dispatch
+    //! between them.
+    class RpcForwarder;
+    std::map<Ref<const std::string>, RpcForwarder> m_rpc_forwarders;
+};
+
+//! Forwarder for one RPC method.
+class ChainImpl::RpcForwarder
+{
+public:
+    RpcForwarder(const CRPCCommand& command) : m_command(command)
+    {
+        m_command.actor = [this](const JSONRPCRequest& request) { return forwardRequest(request); };
+    }
+
+    bool registerForwarder() { return ::tableRPC.appendCommand(m_command.name, &m_command); }
+
+    void addCommand(const CRPCCommand& command) { m_commands.emplace_back(&command); }
+
+    void removeCommand(const CRPCCommand& command)
+    {
+        m_commands.erase(std::remove(m_commands.begin(), m_commands.end(), &command));
+    }
+
+    UniValue forwardRequest(const JSONRPCRequest& request) const
+    {
+        // Simple forwarding of RPC requests. This just sends the request to the
```

Comment:
Supporting multiple clients makes this more complex than it needs to be.

This PR would be easier to review if it initially supported just a single client (as we do today). A future PR could extend the RPC forwarding to support multiple clients.